File: backend/app/api/tools.py
# Built with Spec4 AI - https://spec4.ai
import structlog
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.db.session import get_db_session
from backend.app.services import text_gate
from backend.app.services.moderation import Moderator, get_moderator
from backend.app.tools.examples import EXAMPLE_QUERIES
from backend.app.tools.schemas import (
    AgentStepOut,
    SearchRequest,
    SearchResponse,
    SearchResultOut,
)
from backend.app.tools.service import ToolServiceError, run_search

logger = structlog.get_logger()

router = APIRouter()

#: Tag on this app's moderation-log rows.
TOOLS_APP_NAME = "Tool-Use Example App"


@router.post("/api/tools/search")
async def search_tool(
    payload: SearchRequest,
    session: AsyncSession = Depends(get_db_session),
    moderator: Moderator = Depends(get_moderator),
) -> JSONResponse:
    """Run the tool-use agent loop and return its answer, trace, and results.

    Google-style docstring per project convention.

    Args:
        payload: The visitor's question. The agent decides for itself whether
            and what to search -- this is not passed to the search API as-is.
        session: An async DB session injected per-request.

    Returns:
        A 200 JSON response with the agent's answer, the step trace, and the
        ranked results it saw, or 503 with a clear 'temporarily unavailable'
        message if a usage cap is reached, the Exa Search API call fails, or
        the free-model chain is exhausted.
    """
    gate = await text_gate.check_free_text(
        payload.search_query,
        app_name=TOOLS_APP_NAME,
        curated=EXAMPLE_QUERIES,
        session=session,
        moderator=moderator,
    )
    if not gate.allowed and gate.code is not None:
        return JSONResponse(
            status_code=text_gate.status_for(gate.code),
            content={"status": "error", "code": gate.code, "detail": gate.message},
        )

    try:
        run = await run_search(session, payload.search_query)
    except ToolServiceError as exc:
        logger.error("tool_search_failed", error=str(exc))
        return JSONResponse(
            status_code=503,
            content={"status": "error", "detail": str(exc)},
        )

    logger.info(
        "tool_search_succeeded",
        result_count=len(run.results),
        searches=len(run.queries),
        iterations=run.iterations,
        model=run.model,
    )

    response = SearchResponse(
        answer=run.answer,
        results=[
            SearchResultOut(
                title=r.title, summary=r.summary, source=r.source, rank=r.rank
            )
            for r in run.results
        ],
        steps=[
            AgentStepOut(kind=s.kind, label=s.label, detail=s.detail) for s in run.steps
        ],
        queries=run.queries,
        model=run.model,
        iterations=run.iterations,
    )
    return JSONResponse(status_code=200, content=response.model_dump())

File: frontend/src/api/tools.ts
// Built with Spec4 AI - https://spec4.ai
export interface SearchResult {
  title: string
  summary: string
  source: string
  rank: number
  /**
   * When the page was published, as Exa reports it, or null when it could not
   * determine one.
   *
   * Shown on the card because **relevance is not recency**: Exa ranks purely
   * on relevance, so a 2023 article can out-rank a current one for a
   * time-sensitive question. Without a date on screen, a stale *page* is
   * indistinguishable from the demo answering out of stale training data —
   * which is exactly how it was read.
   */
  published_date: string | null
}

/**
 * One observable step from the backend agent loop. These reflect what the
 * model actually did — which tool it chose, what query it wrote, what came
 * back — rather than a cosmetic animation.
 */
export interface AgentStep {
  kind: 'decision' | 'tool_call' | 'tool_result' | 'answer'
  label: string
  detail: string
}

export interface SearchResponse {
  answer: string
  results: SearchResult[]
  steps: AgentStep[]
  queries: string[]
  model: string
  iterations: number
}

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000'

/**
 * Run the tool-use agent loop over a visitor's question.
 *
 * The question is not sent to the search API as-is: the backend hands the
 * model a `web_search` tool schema, and the model decides whether to search
 * and writes its own query.
 *
 * @param searchQuery - The visitor's question.
 * @returns The agent's answer, its step trace, and the results it saw.
 * @throws Error if the backend responds with a non-2xx status (e.g. a usage
 * cap has been reached or the free-model chain is exhausted, per the
 * tool_use_integration failure mode).
 */
export async function searchTool(searchQuery: string): Promise<SearchResponse> {
  const response = await fetch(`${API_BASE_URL}/api/tools/search`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ search_query: searchQuery }),
  })

  const body = (await response.json()) as SearchResponse | { status: string; detail: string }

  if (!response.ok) {
    throw new Error(
      'detail' in body ? body.detail : `Searching failed with status ${response.status}`,
    )
  }

  return body as SearchResponse
}
